### ISSUE TITLE
Double mutex locking with CURL_LOCK_DATA_CONNECT

### ISSUE THREAD

**Issue by ahodesuka** [curl/curl#2151]
### I did this
I don't know if it's related to #2132, but I am experiencing a double lock while using `CURL_LOCK_DATA_CONNECT` with the multi interface.

Callstack is the following:
```
#0  0x00007ffff0ba259c in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007ffff0b9a45d in pthread_mutex_lock () from /lib64/libpthread.so.0
#2  0x00000000004439ad in __gthread_mutex_lock (__mutex=<optimized out>) at /usr/lib/gcc/x86_64-pc-linux-gnu/6.4.0/include/g++-v6/x86_64-pc-linux-gnu/bits/gthr-default.h:748
#3  std::mutex::lock (this=<optimized out>) at /usr/lib/gcc/x86_64-pc-linux-gnu/6.4.0/include/g++-v6/bits/std_mutex.h:103
#4  AhoViewer::Booru::Site::share_lock_cb (data=<optimized out>, userp=0xe3e9d0) at booru/site.cc:131
#5  0x00007ffff7b5952b in Curl_share_lock (data=0x21e0f68, type=CURL_LOCK_DATA_CONNECT, accesstype=CURL_LOCK_ACCESS_SINGLE) at share.c:227
#6  0x00007ffff7b77ab9 in Curl_conncache_remove_conn (connc=0xe3ee18, conn=0x7fffd4006648) at conncache.c:298
#7  0x00007ffff7b32f63 in Curl_disconnect (conn=0x7fffd4006648, dead_connection=true) at url.c:780
#8  0x00007ffff7b3355d in disconnect_if_dead (conn=0x7fffd4006648, data=0x21e0f68) at url.c:1022
#9  0x00007ffff7b33596 in call_disconnect_if_dead (conn=0x7fffd4006648, param=0x21e0f68) at url.c:1038
#10 0x00007ffff7b77cec in Curl_conncache_foreach (data=0x21e0f68, connc=0xe3ee18, param=0x21e0f68, func=0x7ffff7b3356b <call_disconnect_if_dead>) at conncache.c:348
#11 0x00007ffff7b3360c in prune_dead_connections (data=0x21e0f68) at url.c:1053
#12 0x00007ffff7b3a7fc in create_conn (data=0x21e0f68, in_connect=0x21e0f78, async=0x7fffc5ffa987) at url.c:4397
#13 0x00007ffff7b3b03c in Curl_connect (data=0x21e0f68, in_connect=0x21e0f78, asyncp=0x7fffc5ffa987, protocol_done=0x7fffc5ffa988) at url.c:4687
#14 0x00007ffff7b53fdc in multi_runsingle (multi=0x1768018, now=..., data=0x21e0f68) at multi.c:1432
#15 0x00007ffff7b5649c in multi_socket (multi=0x1768018, checkall=false, s=-1, ev_bitmask=0, running_handles=0x1767e68) at multi.c:2619
#16 0x00007ffff7b56cea in curl_multi_socket_action (multi=0x1768018, s=-1, ev_bitmask=0, running_handles=0x1767e68) at multi.c:2726
#17 0x0000000000430ac8 in AhoViewer::Booru::ImageFetcher::timeout_cb (this=0x1767e10) at booru/imagefetcher.cc:180
```

Following the callstack I see that `Curl_conncache_foreach` is locking, and then `Curl_conncache_remove_conn` is as well.  Using a recursive mutex would solve the problem on my end but that does not seem like the intention of libcurl.
### I expected the following

For everything to just werk

### curl/libcurl version

7.57.0, 7.57.1-DEV, 7.57.1-DEV (bagder/conncache-lockmore)
Tried all 3.

### operating system

Gentoo Linux

**Comment by bagder** [comment]
ack, I see it in the code...

**Comment by bagder** [comment]
It is separate from #2132, but as I'm doing some connection cache locking cleanups in the #2139 PR, I have a fix pending in there for this #2151 issue as well.

**Comment by bagder** [comment]
There, I've now pushed my double-lock fix as part of the general cleanup in the #2139 PR. Would be great if you could see if it fixes your issue.

**Comment by ahodesuka** [comment]
Seems to be functioning correctly now!
Thanks for the quick work.
